druid/src/widget/immediate.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{widget::prelude::*, Data};
+
+/// A widget that allows writing parts of the ui in an immediate-mode style.
+///
+/// The primary use case for this is displaying data that doesn't fit into druid's model, such as
+/// enums you have no control over.
+///
+/// Whenever the state represented by the `Immediate` changes,
+/// it will reconstruct its content for the new data.
+///
+/// While this is not the most efficient thing to do, it is very simple and performs perfectly fine
+/// for small or rarely changed data.
+///
+/// You should only use `Immediate` if your data format can't be reasonably used with other widgets.
+pub struct Immediate<D, W: Widget<()>> {
+    constructor: Box<dyn Fn(&D) -> W>,
+    content: Option<W>,
+}
+
+impl<D, W: Widget<()>> Immediate<D, W> {
+    /// Takes a constructor for a stateless widget
+    pub fn new(constructor: impl Fn(&D) -> W + 'static) -> Self {
+        Self {
+            constructor: Box::new(constructor),
+            content: None,
+        }
+    }
+}
+
+impl<D: Data, W: Widget<()>> Widget<D> for Immediate<D, W> {
```

Comment:
I think the inner widget should have the same `Data` as self, and should use a lens to determine what its inner data type is? I appreciate this is a sort of academic point since you're explicitly trying to not have the inner type use data? But it seems like an unnecessary limitation.

Reply:
I chose `()` explicitly because it prevents one from using anything that depends on data, such as Label::dynamic. Using Label::dynamic does not make any sens in immediate-mode as the Label will be reconstructed anyway when the data changes. Using this widget means giving up the advantages druid offers, this should be clear.